### PR TITLE
Introduce observer: a plugin that pushes project card status to BigQuery

### PR DESCRIPTION
**Pull request by csweichel**
Is already deployed in the prow cluster and happily pumping data:
image

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitbot/pull/17#" title="Author self-approved">csweichel</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitbot).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitbot/blob/main/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: plugins/observer/main.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+type Config struct {
+	BigQuery struct {
+		CredentialFile string `json:"credentials"`
+		ProjectID      string `json:"projectID"`
+		Dataset        string `json:"dataset"`
+	} `json:"bigquery"`
+	OrgRepos map[string]ManagedOrgRepo `json:"orgsRepos,omitempty"`
+}
+
+// ManagedOrgRepo is used by the ObserverC plugin to represent an Organisation
```
Comment (leodido) [inline plugins/observer/main.go:41]:
```suggestion
// ManagedOrgRepo is used by the Observer plugin to represent an Organisation
```

Typo?

File: plugins/observer/main.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+type ManagedOrgRepo struct {
+	Projects map[int]ManagedProject `json:"projects,omitempty"`
+	// with a list of Columns
+}
+
+// ManagedProject is used by the ObserverC plugin to represent a Project
```
Comment (leodido) [inline plugins/observer/main.go:47]:
```suggestion
// ManagedProject is used by the Observer plugin to represent a Project
```

Typo?

**Comment by ArthurSens** [comment]
That looks so nice! Are you planning some statistics analyzes to how quickly issues go from created to done?

**Review by leodido — COMMENTED** [review]
Where's the definition of it as an external plugin?

I suspect there's a missing piece, namely the configuration under "external_plugins" key

File: plugins/observer/main.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+				Nodes []struct {
+					Name  string
+					Cards struct {
+						Nodes []struct {
+							Content struct {
+								PR    queryIssue `graphql:"... on PullRequest"`
+								Issue queryIssue `graphql:"... on Issue"`
+							}
+						}
+					} `graphql:"cards(first: 100)"`
```
Comment (JanKoehnlein) [inline plugins/observer/main.go:250]: Any reason to limit to 100 cards? The „Done“ column could be quite full.

**Review by JanKoehnlein — APPROVED** [review]
Awesome!

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 7fa6b60484f7f2e99c9fdda37c92e696a7182df2</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitbot/pull/17#" title="Author self-approved">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitbot/pull/17#pullrequestreview-717328006" title="Approved">JanKoehnlein</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitbot).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitbot/blob/main/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: plugins/observer/main.go (hunk 3 above)
Comment (csweichel) [reply plugins/observer/main.go:250]: the done column is in the ignore list :)

Comment (csweichel) [reply plugins/observer/main.go:250]: basically once the cards drop out of the query they're considered done if they were in "Await Deployment" before

File: plugins/observer/main.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+					issue = crd.Content.Issue
+				} else if crd.Content.PR.Title != "" {
+					issue = crd.Content.PR
+				} else {
+					continue
+				}
+
+				assignees := make([]string, len(issue.Assignees.Nodes))
+				for i := range issue.Assignees.Nodes {
+					assignees[i] = string(issue.Assignees.Nodes[i].Login)
```
Comment (JanKoehnlein) [inline plugins/observer/main.go:308]: Would it make sense to distinguish between issue assignees and PR assignees (reviewers) for the analysis?

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitbot/pull/17#pullrequestreview-717328006" title="Approved">JanKoehnlein</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **csweichel** after the PR has been reviewed.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitbot).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitbot/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitbot/pull/17#issuecomment-892456098" title="Approved">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitbot/pull/17#pullrequestreview-717328006" title="Approved">JanKoehnlein</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitbot).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitbot/blob/main/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**